### DISCUSSION TITLE
How to scroll ListBox to the last item automatically

### DISCUSSION THREAD

**Discussion by pzs-scnu** [AvaloniaUI/Avalonia#16898] [Q&A] [answered]
I define a ListBox, its ItemsSource is bound to an  ObservableCollection property in ViewModel class, items in this ListBox are continuously added as the app runs. As more and more items are added, some ListBox lines are invisible, now I want to scroll the ListBox to the last item so it is visible in Window,  how to achieve this?

**Comment by workgroupengineering** [comment, marked as answer]
1. set `AutoScrollToSelectedItem="true" 
2. Binding ListBox.SelectedItem on your vm property (eg: `MySelectedItem`)
3. In method were populate  ObservableCollection  set MySelectedItem to last item added, ensure to invoke assignment of in Dispatcher.Post like this:
   ```c#
   var itemsToAdd = ...
   _items.AddRange(itemsToAdd);
   Dispatcher.UIThread.Post(item=>
   {
         MySelectedItem = item as ItemViewModel;
   }
   , itemsToAdd[itemsToAdd.Count-1]
   ,  DispatcherPriority.Background
   ```

**Comment by sora324** [comment]
Any better way to achieve this?

**Reply by stevemonaco** [reply]
`<ListBox x:Name="listBox" />`
```cs
public partial class SandboxPageView : UserControl
{
    private int i = 1;
    
    public SandboxPageView()
    {
        InitializeComponent();

        listBox.ItemsView.CollectionChanged += (sender, args) =>
        {
            if (args.Action != NotifyCollectionChangedAction.Add)
                return;

            Dispatcher.UIThread.Post(() =>
            {
                var sv = listBox.Scroll as ScrollViewer;
                sv?.ScrollToEnd();
            });
        };

        DispatcherTimer.Run(() =>
        {
            listBox.Items.Add(i);
            i++;
            return i <= 200;
        }, TimeSpan.FromMilliseconds(50));
    }
}
```

Using the `Dispatcher` to post the `ScrollToEnd` is important here as the `ListBox` UI changes from the collection change are still in flight. It's better to monitor `CollectionChanged` here than `ContainerPrepared` event.

**Reply by sora324** [reply]
Thanks.
